### PR TITLE
Laboratory work #4, Valeria Bogdanova - 23FPL1

### PR DESCRIPTION
**Pull request by lerabogdanova**

### CODE REVIEW

File: lab_4_retrieval_w_clustering/main.py [hunk 1]
```diff
@@ -67,6 +81,13 @@ def vectorize(self, tokenized_document: list[str]) -> Vector:
         Returns:
             Vector: BM25 vector for document.
         """
+        if not tokenized_document or not isinstance(tokenized_document, list) \
+            or not all(isinstance(token, str) for token in tokenized_document):
+            raise ValueError
+        if self._calculate_bm25(tokenized_document) is None:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:88]: add message

File: lab_4_retrieval_w_clustering/main.py [hunk 2]
```diff
@@ -67,6 +81,13 @@ def vectorize(self, tokenized_document: list[str]) -> Vector:
         Returns:
             Vector: BM25 vector for document.
         """
+        if not tokenized_document or not isinstance(tokenized_document, list) \
+            or not all(isinstance(token, str) for token in tokenized_document):
+            raise ValueError
+        if self._calculate_bm25(tokenized_document) is None:
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:87]: You call the same method twice. Why?

File: lab_4_retrieval_w_clustering/main.py [hunk 3]
```diff
@@ -81,6 +102,19 @@ def _calculate_bm25(self, tokenized_document: list[str]) -> Vector:
...
+        if not tokenized_document or not isinstance(tokenized_document, list) \
+            or not all(isinstance(token, str) for token in tokenized_document):
+            raise ValueError
+
+        vector = [0.0 for _ in self._vocabulary]
+        bm25 = calculate_bm25(self._vocabulary, tokenized_document,
+                              self._idf_values,
+                              avg_doc_len=self._avg_doc_len, doc_len=len(tokenized_document))
+        if not bm25:
+            return()
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:114]: Is return a function? Do you call return? what is it? Does it work?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_retrieval_w_clustering/main.py (hunk 2 above)
Comment (lerabogdanova) [reply lab_4_retrieval_w_clustering/main.py:87]: First for None check, separately - not to create a variable

File: lab_4_retrieval_w_clustering/main.py (hunk 3 above)
Comment (lerabogdanova) [reply lab_4_retrieval_w_clustering/main.py:114]: Forgot to change this.... but it worked

File: lab_4_retrieval_w_clustering/main.py [hunk 4]
```diff
@@ -26,6 +28,10 @@ def get_paragraphs(text: str) -> list[str]:
     Returns:
         list[str]: Paragraphs from document.
     """
+    if not text or not isinstance(text, str):
+        raise ValueError
```
[the remark this answers is not in the document]
Comment (lerabogdanova) [reply lab_4_retrieval_w_clustering/main.py:32]: What kind of message? "Bad input"?

File: lab_4_retrieval_w_clustering/main.py [hunk 5]
```diff
@@ -40,6 +46,9 @@ def __init__(self) -> None:
         """
         Initialize an instance of the BM25Vectorizer class.
         """
+        self._corpus = []
+        super().__init__(self._corpus)
```
[the remark this answers is not in the document]
Comment (lerabogdanova) [reply lab_4_retrieval_w_clustering/main.py:50]: It is needed to initialise the 'corpus' attribute but also call super to inherit class with 'corpus' argument in init
So when typing super().__init__(... Python asks for corpus parameter
I don't know how to do it another way ;(
Or I didn't understand the comment..?

File: lab_4_retrieval_w_clustering/main.py (hunk 2 above)
Comment (artyomtugaryov) [reply lab_4_retrieval_w_clustering/main.py]: Really? I clearly understand it. But I suggest saving the result to variable and call this method once

File: lab_4_retrieval_w_clustering/main.py (hunk 3 above)
Comment (artyomtugaryov) [reply lab_4_retrieval_w_clustering/main.py]: Done?

File: lab_4_retrieval_w_clustering/settings.json [hunk 6]
```diff
@@ -1,3 +1,3 @@
 {
-    "target_score": 0
+    "target_score": 4
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/settings.json]: You must just send a funny meme for 4.

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_retrieval_w_clustering/main.py (hunk 3 above)
Comment (lerabogdanova) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 1 above)
Comment (lerabogdanova) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py [hunk 7]
```diff
@@ -67,6 +81,13 @@ def vectorize(self, tokenized_document: list[str]) -> Vector:
         Returns:
             Vector: BM25 vector for document.
         """
+        if not tokenized_document or not isinstance(tokenized_document, list) \
+            or not all(isinstance(token, str) for token in tokenized_document):
+            raise ValueError
```
[the remark this answers is not in the document]
Comment (lerabogdanova) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py [hunk 8]
```diff
@@ -51,6 +60,11 @@ def set_tokenized_corpus(self, tokenized_corpus: TokenizedCorpus) -> None:
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not tokenized_corpus:
+            raise ValueError
```
[the remark this answers is not in the document]
Comment (lerabogdanova) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lerabogdanova) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py [hunk 9]
```diff
@@ -170,6 +235,8 @@ def __init__(self, db: DocumentVectorDB) -> None:
         Args:
             db (DocumentVectorDB): Object of DocumentVectorDB class.
         """
+        self._db = db
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:243]: move it after `super`

File: lab_4_retrieval_w_clustering/main.py [hunk 10]
```diff
@@ -81,6 +103,19 @@ def _calculate_bm25(self, tokenized_document: list[str]) -> Vector:
         Returns:
             Vector: BM25 vector for document.
         """
+        if not tokenized_document or not isinstance(tokenized_document, list) \
+            or not all(isinstance(token, str) for token in tokenized_document):
+            raise ValueError("Bad input")
+
+        vector = [0.0 for _ in self._vocabulary]
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: try to create a list from one item and then multiply it by the length of vocabulary

File: lab_4_retrieval_w_clustering/main.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (artyomtugaryov) [reply lab_4_retrieval_w_clustering/main.py]: You can call super with empty list and the create corpus, right?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_retrieval_w_clustering/main.py (hunk 10 above)
Comment (lerabogdanova) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 2 above)
Comment (lerabogdanova) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lerabogdanova) [reply lab_4_retrieval_w_clustering/main.py]: Done

**Pull request closed without merging**